accepted/2023/dotnet-nuget-setapikey.md

```diff
@@ -0,0 +1,230 @@
+
+## Rationale and alternatives
+
+### Which NuGet.config file
+
+At the time this spec is written, all NuGet commands that can write values in `NuGet.config` files have the following unintuitive behavior.
+The docs on [common NuGet configurations](https://learn.microsoft.com/nuget/consume-packages/configuring-nuget-behavior) explains that there's a hierarchy of directories and files checked.
+When making a change, NuGet will look for the first file in that ordered list that contains the relevant section (first child of the XML root element).
+For example, `nuget.exe config` will look for the first file that contains `<config>`, and `dotnet nuget add source` will look for the first file that contains `<packageSources>`.
+While package sources, and certainly other config settings are valuable to be in a solution `NuGet.config` file, and therefore commit into source control, accidentally committing a secret into source control is a security risk.
+Therefore, `dotnet nuget apikey set` will always default to the user scope `NuGet.config` file.
+
+On stateful CI agents, this poses a security risk.
+If the pipeline does not remove the API key from the user nuget.config, then other pipelines running on the same agent will be able to read the API key from the config file.
+However, a pipeline that fails to clean up after itself on a stateful CI agent (for example times out, or the agent crashes or restarts without waiting for the pipeline to finish) will similarly leave secrets that other pipelines can read.
+Although, if the secret is in a repo-specific directory, it will be less discoverable by other pipelines.
+But a balance needs to be made between defaults for devboxes or CI agents, since the preferred default for each scenario is different.
+If it were possible to easily detect when NuGet is running on a CI agent, it might be good to detect and change the defaults.
+However, [there is no standardization across CI platforms](https://adamj.eu/tech/2020/03/09/detect-if-your-tests-are-running-on-ci/).
+
+<!-- Why is this the best design compared to other designs? -->
+<!-- What other designs have been considered and why weren't they chosen? -->
+<!-- What is the impact of not doing this? -->
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevant to this proposal? -->
+
+- `dotnet nuget apikey` commands `set` and `unset`
+
+The [`dotnet nuget config` spec](https://github.com/NuGet/Home/pull/12172) proposed `set` and `unset`, to mimic `git config` commands, rather than `add` and `remove` or `delete`, which commands like `dotnet add package` and `dotnet nuget add source` use.
```

Comment:
FYI there are `dotnet user-secrets set` and `dotnet user-secrets remove` commands (also git also has `git remote remove`). If you haven't already, consider double-checking what's considered idiomatic by the .NET tool team.

Reply:
I do like set/unset over add/remove. `add` IMHO is ambiguous regarding what happens when you attempt to overwrite. I.e., Will it overwrite the old value? Will it ignore the new value because a value is already present? Will it return an error because a value is already present?

Reply:
I think in the scope of this command where we are adding a new apikeys element to a config, you could use the `add/remove` consistency as seen in other commands.

I know it isn't always representative that it can be overwritten (i.e. dotnet add package), but it is the term that was chosen and would be remembered more easily by CLI users who use all other CRUD operations. 

My second vote would be `set / remove`.

I personally see this command as a means to "add a apikey to a source". Similar to if you were to just add the apikey in the `dotnet nuget add source --apikey <blah>` command (if it was supported)